python/tritonserver/_api/_model.py

```diff
@@ -0,0 +1,491 @@
+from typing_extensions import Unpack
+
+if TYPE_CHECKING:
+    from tritonserver._api._server import Server
+
+
+class Model:
+    """Class for interacting with Triton Inference Server models
+
+    Model objects are returned from server factory methods and allow
+    users to query metadata and execute inference
+    requests.
+
+    """
+
+    def __init__(
+        self,
+        server: Server,
+        name: str,
+        version: int = -1,
+        state: Optional[str] = None,
+        reason: Optional[str] = None,
+    ) -> None:
+        """Initialize model
+
+        Model objects should be obtainted from Server factory methods
+        and not instantiated directly. See `Server` documentation.
+
+        Parameters
+        ----------
+        server : Server
+            Server instance.
+        name : str
+            model name
+        version : int
+            model version
+        state : Optional[str]
+            state of model (if known)
+        reason : Optional[str]
+            reason for model state (if known)
+
+        Examples
+        --------
+        >>> server.model("test")
+        server.model("test")
+        {'name': 'test', 'version': -1, 'state': None}
+
+        """
+
+        self._server = server._server
+        if not isinstance(self._server, TRITONSERVER_Server):
+            raise InvalidArgumentError("Server not started")
+        self.name = name
+        self.version = version
+        self.state = state
+        self.reason = reason
+
+    def create_request(self, **kwargs: Unpack[InferenceRequest]) -> InferenceRequest:
+        """Inference request factory method
+
+        Return an inference request object that can be used with
+        model.infer() ro model.async_infer()
+
+        Parameters
+        ----------
+
+        kwargs : Unpack[InferenceRequest]
+            Keyword arguments passed to `InferenceRequest` constructor. See
+            `InferenceRequest` documentation for details.
+
+        Returns
+        -------
+        InferenceRequest
+            Inference request associated with this model
+
+        Examples
+        --------
+
+        >>> server.model("test").create_request()
+        server.model("test").create_request()
+        InferenceRequest(model={'name': 'test', 'version': -1,
+        'state': None},
+        _server=<tritonserver._c.triton_bindings.TRITONSERVER_Server
+        object at 0x7f5827156bf0>, request_id=None, flags=0,
+        correlation_id=None, priority=0, timeout=0, inputs={},
+        parameters={}, output_memory_type=None,
+        output_memory_allocator=None, response_queue=None,
+        _serialized_inputs={})
+
+        """
+        if "model" in kwargs:
+            kwargs.pop("model")
+        return InferenceRequest(model=self, **kwargs)
+
+    def async_infer(
+        self,
+        inference_request: Optional[InferenceRequest] = None,
+        raise_on_error: bool = False,
```

Comment:
I think probably the pythonic way is to always raise an exception on error.

cc @rmccorm4 in case you have any thoughts here.

Reply:
how about switch default? I think its useful in case of multiple responses where some may have error and some may not - "raise_on_error" is relatively well used (open telemetry, elasticsearch) for different scenarios

Reply:
sounds good.

Reply:
LGTM